File: legacy-python/src/attitude_sim/reaction_wheels.py
"""Three-axis reaction-wheel assembly for the SimLab actuator path.

This is an **additive** actuator model.  It does not rewrite
``attitude_sim.gyrostat`` (one-rotor dual-spin plant), ``polhode``, or
``RigidBody``.  Closed-loop SimLab still integrates Euler's equation on
the rigid body; the wheels contribute the applied torque that already
enters after the PID/LQR command (the existing ``TorqueActuator`` sat /
lag path).

Orthogonal body-aligned wheels with diagonal inertia ``I_w``:

    h_w = I_w Ω
    I_w Ω̇ = τ_m + τ_f + τ_dump

The motor torque is the reaction of the body command after per-axis
``|τ|`` saturation and a momentum-saturation gate

    τ_m^des = −τ_cmd
    τ_m = clip(τ_m^des, ±τ_max)

and if ``|h_i| ≥ h_max,i`` and ``τ_m,i h_i > 0`` then ``τ_m,i ← 0``
(cannot spin further into the wall).  Viscous + smoothed Coulomb
friction on the wheel is

    τ_f = −b Ω − c tanh(Ω / ε)

The torque passed to ``RigidBody`` (which already includes ``−ω × Jω``) is

    τ = −ḣ_w − ω × h_w

so the closed-loop plant matches the gyrostat Euler equation without
calling ``Gyrostat`` / ``step_gyrostat``.  ``gyroscopic=False`` drops
the ``ω × h_w`` term.

``τ_dump`` is a reserved wheel-side dump torque for a future
magnetorquer coupler.  Magnetorquers are **not** modeled here — do not
duplicate a mag-torquer module in this file.

Default construction with no ``I_w`` / ``h_max`` / friction is not
used; ``make_actuator`` still returns identity ``TorqueActuator`` so
prior closed-loop runs match.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from attitude_sim.actuators import clip_torque

DEFAULT_WHEEL_INERTIA = 2.0e-4
"""Smallsat-class spin inertia per axis (kg·m²) when only ``h_max`` is set."""

_COULOMB_EPS = 0.05  # rad/s; smooth sign so Coulomb does not chatter at rest
_SAT_FRAC = 1.0 - 1e-9


def _as_positive_axes(
    value: float | np.ndarray,
    name: str,
) -> np.ndarray:
    """Broadcast a positive scalar or length-3 vector."""
    arr = np.asarray(value, dtype=float)
    if arr.ndim == 0:
        out = np.full(3, float(arr))
    else:
        out = arr.reshape(3).copy()
    if not np.all(np.isfinite(out)):
        raise ValueError(f"{name} must be finite")
    if np.any(out <= 0.0):
        raise ValueError(f"{name} must be strictly positive, got {out}")
    return out


def _as_nonneg_axes(
    value: float | np.ndarray | None,
    name: str,
) -> np.ndarray | None:
    if value is None:
        return None
    # Reuse clip_torque validation (allows a zero limit that locks an axis).
    clip_torque(np.zeros(3), value)
    arr = np.asarray(value, dtype=float)
    if arr.ndim == 0:
        return np.full(3, float(arr))
    return arr.reshape(3).copy()


def gate_momentum_saturation(
    tau_m: np.ndarray,
    h: np.ndarray,
    h_max: float | np.ndarray | None,
) -> np.ndarray:
    """Zero motor torque that would drive |h_i| further past h_max,i."""
    out = np.asarray(tau_m, dtype=float).reshape(3).copy()
    if h_max is None:
        return out
    lim = _as_nonneg_axes(h_max, "h_max")
    if lim is None:  # pragma: no cover
        return out
    h = np.asarray(h, dtype=float).reshape(3)
    at_hi = h >= lim * _SAT_FRAC
    at_lo = h <= -lim * _SAT_FRAC
    out = np.where(at_hi & (out > 0.0), 0.0, out)
    out = np.where(at_lo & (out < 0.0), 0.0, out)
    return out


def friction_torque(
    omega_w: np.ndarray,
    visc: float,
    coulomb: float,
    *,
    eps: float = _COULOMB_EPS,
) -> np.ndarray:
    """Wheel-frame friction τ_f = −b Ω − c tanh(Ω / ε)."""
    w = np.asarray(omega_w, dtype=float).reshape(3)
    b = float(visc)
    c = float(coulomb)
    if b < 0.0 or c < 0.0:
        raise ValueError(f"friction coefficients must be >= 0, got visc={b}, coulomb={c}")
    if not np.isfinite(eps) or eps <= 0.0:
        raise ValueError("coulomb smoothing eps must be positive")
    return -b * w - c * np.tanh(w / eps)


@dataclass
class ReactionWheelAssembly:
    """Three-axis RW: τ_cmd → ḣ_w with |τ|, |h| sat and friction.

    ``apply`` has the same contract as :class:`~attitude_sim.actuators.TorqueActuator`
    plus an optional body rate for the gyroscopic ω × h_w couple.
    """

    tau_max: float | np.ndarray | None = None
    h_max: float | np.ndarray | None = None
    wheel_inertia: float | np.ndarray = DEFAULT_WHEEL_INERTIA
    visc_friction: float = 0.0
    coulomb_friction: float = 0.0
    time_constant: float | None = None
    gyroscopic: bool = True
    coulomb_eps: float = _COULOMB_EPS
    # Reserved magnetorquer dump torque on the wheels (N·m).  External to
    # this module; SimLab does not synthesize an MTQ law here.
    tau_dump: np.ndarray | None = None
    _h: np.ndarray = field(default_factory=lambda: np.zeros(3), init=False, repr=False)
    _tau_m: np.ndarray = field(default_factory=lambda: np.zeros(3), init=False, repr=False)
    _tau: np.ndarray = field(default_factory=lambda: np.zeros(3), init=False, repr=False)
    _tau_sat: np.ndarray = field(
        default_factory=lambda: np.zeros(3, dtype=bool), init=False, repr=False
    )
    _h_sat: np.ndarray = field(
        default_factory=lambda: np.zeros(3, dtype=bool), init=False, repr=False
    )
    _I_w: np.ndarray = field(init=False, repr=False)
    _h_lim: np.ndarray | None = field(init=False, repr=False)

    def __post_init__(self) -> None:
        if self.time_constant is not None and float(self.time_constant) < 0.0:
            raise ValueError(f"actuator time constant must be >= 0, got {self.time_constant}")
        self._I_w = _as_positive_axes(self.wheel_inertia, "wheel_inertia")
        self._h_lim = _as_nonneg_axes(self.h_max, "h_max")
        if self.tau_max is not None:
            clip_torque(np.zeros(3), self.tau_max)
        if float(self.visc_friction) < 0.0 or float(self.coulomb_friction) < 0.0:
            raise ValueError("friction coefficients must be >= 0")
        if self.tau_dump is not None:
            self.tau_dump = np.asarray(self.tau_dump, dtype=float).reshape(3).copy()

    def reset(
        self,
        tau0: np.ndarray | None = None,
        h0: np.ndarray | None = None,
    ) -> None:
        """Zero (or set) stored momentum and lagged motor torque."""
        del tau0  # body torque is not a wheel state; keep TorqueActuator signature
        if h0 is None:
            self._h[:] = 0.0
        else:
            h = np.asarray(h0, dtype=float).reshape(3)
            self._h[:] = clip_torque(h, self.h_max) if self.h_max is not None else h
        self._tau_m[:] = 0.0
        self._tau[:] = 0.0
        self._tau_sat[:] = False
        self._h_sat[:] = False

    @property
    def inertia(self) -> np.ndarray:
        """Wheel spin inertias I_w (kg·m²)."""
        return self._I_w.copy()

    @property
    def momentum(self) -> np.ndarray:
        """Wheel angular momentum h_w (N·m·s)."""
        return self._h.copy()

    @property
    def wheel_speed(self) -> np.ndarray:
        """Relative wheel rates Ω = I_w^{-1} h_w (rad/s)."""
        return self._h / self._I_w

    @property
    def torque(self) -> np.ndarray:
        """Last applied body torque (N·m)."""
        return self._tau.copy()

    @property
    def torque_saturated(self) -> np.ndarray:
        """Per-axis motor-torque saturation flags at the last sample."""
        return self._tau_sat.copy()

    @property
    def momentum_saturated(self) -> np.ndarray:
        """Per-axis momentum-saturation flags at the last sample."""
        return self._h_sat.copy()

    def apply(
        self,
        command: np.ndarray,
        dt: float,
        omega: np.ndarray | None = None,
    ) -> np.ndarray:
        """Advance wheels one sample.  Returns the body torque for ``RigidBody``."""
        dt = float(dt)
        if not np.isfinite(dt) or dt <= 0.0:
            raise ValueError(f"dt must be positive, got {dt}")
        tau_cmd = np.asarray(command, dtype=float).reshape(3)
        tau_m_des = clip_torque(-tau_cmd, self.tau_max)
        T = None if self.time_constant is None else float(self.time_constant)
        if T is None or T <= 0.0:
            tau_m = tau_m_des
        else:
            alpha = float(np.exp(-dt / T))
            tau_m = alpha * self._tau_m + (1.0 - alpha) * tau_m_des
            tau_m = clip_torque(tau_m, self.tau_max)
        tau_m = gate_momentum_saturation(tau_m, self._h, self.h_max)
        omega_w = self._h / self._I_w
        tau_f = friction_torque(
            omega_w,
            self.visc_friction,
            self.coulomb_friction,
            eps=self.coulomb_eps,
        )
        dump = (
            np.zeros(3)
            if self.tau_dump is None
            else np.asarray(self.tau_dump, dtype=float).reshape(3)
        )
        h_dot_unsat = tau_m + tau_f + dump
        h_next = self._h + h_dot_unsat * dt
        if self.h_max is not None:
            h_next = clip_torque(h_next, self.h_max)
        h_dot = (h_next - self._h) / dt
        gyro = np.zeros(3)
        if self.gyroscopic and omega is not None:
            w = np.asarray(omega, dtype=float).reshape(3)
            gyro = np.cross(w, self._h)
        tau_body = -h_dot - gyro
        self._h = h_next
        self._tau_m = tau_m
        self._tau = tau_body
        self._tau_sat = _at_limit(tau_cmd, self.tau_max)
        self._h_sat = _at_limit(self._h, self.h_max)
        return tau_body.copy()


def _at_limit(
    value: np.ndarray,
    limit: float | np.ndarray | None,
) -> np.ndarray:
    """True on axes sitting on a finite per-axis box."""
    if limit is None:
        return np.zeros(3, dtype=bool)
    lim = np.asarray(limit, dtype=float)
    box = np.full(3, float(lim)) if lim.ndim == 0 else lim.reshape(3)
    return np.abs(np.asarray(value, dtype=float).reshape(3)) >= box * _SAT_FRAC


def make_reaction_wheels(
    *,
    tau_max: float | np.ndarray | None = None,
    h_max: float | np.ndarray | None = None,
    wheel_inertia: float | np.ndarray | None = None,
    visc_friction: float = 0.0,
    coulomb_friction: float = 0.0,
    time_constant: float | None = None,
    gyroscopic: bool = True,
) -> ReactionWheelAssembly:
    """Factory matching ``make_actuator`` / ``make_controller`` style."""
    inertia: float | np.ndarray = (
        DEFAULT_WHEEL_INERTIA if wheel_inertia is None else wheel_inertia
    )
    return ReactionWheelAssembly(
        tau_max=tau_max,
        h_max=h_max,
        wheel_inertia=inertia,
        visc_friction=visc_friction,
        coulomb_friction=coulomb_friction,
        time_constant=time_constant,
        gyroscopic=gyroscopic,
    )

File: engine/src/reaction_wheels.rs
//! Three-axis reaction-wheel assembly: motor torque, momentum storage,
//! per-axis |tau|/|h| saturation, viscous + smoothed-Coulomb friction, and
//! the omega x h_w gyroscopic couple. Additive actuator stage sitting after
//! the controller command, matching `attitude_sim.reaction_wheels`.

use crate::actuators::{clip_torque, AxisLimit};
use crate::quaternion::Vec3;

pub const DEFAULT_WHEEL_INERTIA: f64 = 2.0e-4;
const COULOMB_EPS: f64 = 0.05;
const SAT_FRAC: f64 = 1.0 - 1e-9;

fn gate_momentum_saturation(tau_m: &Vec3, h: &Vec3, h_max: Option<AxisLimit>) -> Vec3 {
    match h_max {
        None => *tau_m,
        Some(lim) => {
            let l = lim.as_vec3();
            let mut out = *tau_m;
            for i in 0..3 {
                let at_hi = h[i] >= l[i] * SAT_FRAC;
                let at_lo = h[i] <= -l[i] * SAT_FRAC;
                if (at_hi && out[i] > 0.0) || (at_lo && out[i] < 0.0) {
                    out[i] = 0.0;
                }
            }
            out
        }
    }
}

fn friction_torque(omega_w: &Vec3, visc: f64, coulomb: f64) -> Vec3 {
    Vec3::new(
        -visc * omega_w.x - coulomb * (omega_w.x / COULOMB_EPS).tanh(),
        -visc * omega_w.y - coulomb * (omega_w.y / COULOMB_EPS).tanh(),
        -visc * omega_w.z - coulomb * (omega_w.z / COULOMB_EPS).tanh(),
    )
}

fn at_limit(value: &Vec3, limit: Option<AxisLimit>) -> [bool; 3] {
    match limit {
        None => [false, false, false],
        Some(lim) => {
            let l = lim.as_vec3();
            [
                value.x.abs() >= l.x * SAT_FRAC,
                value.y.abs() >= l.y * SAT_FRAC,
                value.z.abs() >= l.z * SAT_FRAC,
            ]
        }
    }
}

#[derive(Debug, Clone)]
pub struct ReactionWheelAssembly {
    pub tau_max: Option<AxisLimit>,
    pub h_max: Option<AxisLimit>,
    pub wheel_inertia: Vec3,
    pub visc_friction: f64,
    pub coulomb_friction: f64,
    pub time_constant: Option<f64>,
    pub gyroscopic: bool,
    h: Vec3,
    tau_m: Vec3,
    tau: Vec3,
    tau_sat: [bool; 3],
    h_sat: [bool; 3],
}

impl ReactionWheelAssembly {
    pub fn new(
        tau_max: Option<AxisLimit>,
        h_max: Option<AxisLimit>,
        wheel_inertia: Vec3,
        visc_friction: f64,
        coulomb_friction: f64,
        time_constant: Option<f64>,
        gyroscopic: bool,
    ) -> Self {
        ReactionWheelAssembly {
            tau_max,
            h_max,
            wheel_inertia,
            visc_friction,
            coulomb_friction,
            time_constant,
            gyroscopic,
            h: Vec3::zeros(),
            tau_m: Vec3::zeros(),
            tau: Vec3::zeros(),
            tau_sat: [false; 3],
            h_sat: [false; 3],
        }
    }

    pub fn torque(&self) -> Vec3 {
        self.tau
    }

    pub fn momentum(&self) -> Vec3 {
        self.h
    }

    pub fn wheel_speed(&self) -> Vec3 {
        Vec3::new(
            self.h.x / self.wheel_inertia.x,
            self.h.y / self.wheel_inertia.y,
            self.h.z / self.wheel_inertia.z,
        )
    }

    pub fn torque_saturated(&self) -> [bool; 3] {
        self.tau_sat
    }

    pub fn momentum_saturated(&self) -> [bool; 3] {
        self.h_sat
    }

    /// Advance the wheels one sample. Returns the body torque applied to `RigidBody`.
    pub fn apply(&mut self, command: &Vec3, dt: f64, omega: &Vec3) -> Vec3 {
        let tau_m_des = clip_torque(&(-command), self.tau_max);
        let tau_m = match self.time_constant {
            Some(t) if t > 0.0 => {
                let alpha = (-dt / t).exp();
                clip_torque(
                    &(alpha * self.tau_m + (1.0 - alpha) * tau_m_des),
                    self.tau_max,
                )
            }
            _ => tau_m_des,
        };
        let tau_m = gate_momentum_saturation(&tau_m, &self.h, self.h_max);
        let omega_w = self.wheel_speed();
        let tau_f = friction_torque(&omega_w, self.visc_friction, self.coulomb_friction);
        let h_dot_unsat = tau_m + tau_f;
        let mut h_next = self.h + h_dot_unsat * dt;
        if self.h_max.is_some() {
            h_next = clip_torque(&h_next, self.h_max);
        }
        let h_dot = (h_next - self.h) / dt;
        let gyro = if self.gyroscopic {
            omega.cross(&self.h)
        } else {
            Vec3::zeros()
        };
        let tau_body = -h_dot - gyro;

        self.h = h_next;
        self.tau_m = tau_m;
        self.tau = tau_body;
        self.tau_sat = at_limit(command, self.tau_max);
        self.h_sat = at_limit(&self.h, self.h_max);
        tau_body
    }
}

pub fn make_reaction_wheels(
    tau_max: Option<AxisLimit>,
    h_max: Option<AxisLimit>,
    wheel_inertia: Option<AxisLimit>,
    visc_friction: f64,
    coulomb_friction: f64,
    time_constant: Option<f64>,
    gyroscopic: bool,
) -> ReactionWheelAssembly {
    let inertia = wheel_inertia.map(|w| w.as_vec3()).unwrap_or_else(|| {
        Vec3::new(
            DEFAULT_WHEEL_INERTIA,
            DEFAULT_WHEEL_INERTIA,
            DEFAULT_WHEEL_INERTIA,
        )
    });
    ReactionWheelAssembly::new(
        tau_max,
        h_max,
        inertia,
        visc_friction,
        coulomb_friction,
        time_constant,
        gyroscopic,
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_relative_eq;

    #[test]
    fn matches_torque_actuator_when_unlimited_no_gyro() {
        // With gyroscopic=false and no h_max, applied tau should equal -tau_cmd's
        // rate contribution reaction i.e. tracks the box actuator on first sample.
        let mut rw = make_reaction_wheels(None, None, None, 0.0, 0.0, None, false);
        let cmd = Vec3::new(0.01, -0.02, 0.0);
        let tau = rw.apply(&cmd, 0.01, &Vec3::zeros());
        // h_dot = tau_m = -cmd (clip=None) => tau_body = -h_dot = cmd
        assert_relative_eq!(tau, cmd, epsilon = 1e-12);
    }

    #[test]
    fn per_axis_torque_never_exceeds_limit_without_gyroscopic_coupling() {
        // Matches the Python reference's own invariant (`test_reaction_wheels.py`
        // asserts this with `gyroscopic=False`): the *motor* reaction is boxed
        // to tau_max, but with gyroscopic=True the total body torque also
        // carries -omega x h_w, which is a physical consequence of body
        // rotation and is not itself clamped (see the module docs).
        let mut rw = make_reaction_wheels(
            Some(AxisLimit::Scalar(0.02)),
            None,
            None,
            0.0,
            0.0,
            None,
            false,
        );
        let cmd = Vec3::new(1.0, -1.0, 0.5);
        for _ in 0..50 {
            let tau = rw.apply(&cmd, 0.01, &Vec3::new(0.1, -0.1, 0.05));
            assert!(tau.x.abs() <= 0.02 + 1e-9);
            assert!(tau.y.abs() <= 0.02 + 1e-9);
            assert!(tau.z.abs() <= 0.02 + 1e-9);
        }
    }

    #[test]
    fn gyroscopic_coupling_can_exceed_motor_torque_limit() {
        let mut rw = make_reaction_wheels(
            Some(AxisLimit::Scalar(0.02)),
            None,
            None,
            0.0,
            0.0,
            None,
            true,
        );
        let cmd = Vec3::new(1.0, -1.0, 0.5);
        // Build up wheel momentum first so omega x h is non-negligible.
        for _ in 0..2000 {
            rw.apply(&cmd, 0.01, &Vec3::new(0.1, -0.1, 0.05));
        }
        let tau = rw.apply(&cmd, 0.01, &Vec3::new(0.1, -0.1, 0.05));
        assert!(
            tau.norm() > 0.02,
            "expected gyroscopic coupling to push |tau| above tau_max, got {}",
            tau.norm()
        );
    }

    #[test]
    fn momentum_saturates_at_h_max() {
        let mut rw = make_reaction_wheels(
            None,
            Some(AxisLimit::Scalar(0.01)),
            None,
            0.0,
            0.0,
            None,
            false,
        );
        let cmd = Vec3::new(-1.0, 0.0, 0.0); // drives h positive continuously
        for _ in 0..10000 {
            rw.apply(&cmd, 0.01, &Vec3::zeros());
        }
        assert!(rw.momentum().x <= 0.01 + 1e-9);
        assert!(rw.momentum().x >= 0.01 - 1e-6);
    }
}
